lib/puppet/functions/deprecation.rb

```diff
@@ -0,0 +1,28 @@
+# Function to print deprecation warnings, it will display a provided message as well as the calling function.
+# The deprecation message will only be displayed once per caller.
+# It is affected by the puppet setting 'strict', which can be set to :error, :off and :warning (default)
+#
+require "puppet_x/puppetlabs/stdlib/deprecation_state"
+
+Puppet::Functions.create_function(:deprecation) do
+  dispatch :deprecation do
+    param 'String', :message
+  end
+
+  def deprecation(message)
+    called_by = Kernel.caller.first
```

Comment:
We talked about that we want to have the puppet call site (file/line), not the ruby caller (which will always be the same). Maybe @hlindberg knows the magic incantation?